### PR TITLE
Gui refactored into separate views | rewrite of keydown handler into event handler

### PR DESCRIPTION
**Pull request by KyleKlus**
### Description
This PR implements a refactor of the flashcard-modal.tsx into serveral view files (DecksListView, FlashcardReviewView).
This was done to ensure code maintainability for possible later overhauls (eg. issue #872). Like this every view has its separate file, where only the code for this view is put.

Another thing, which this PR implements is the rewritten keydown handler. Now it is written as proper event listener, with extensibility (Through switch statements) and readability in mind.

### Changes

-   Improved how spacing between the buttons was done
-   Centered the button groups, which wheren't centered
-   Fixed a hack in the response of the cram card mode
-   Refactored each view into it's own file, to reduce clutter and having a monolithic file
-   Fixed keydown handler with a proper event handler and a switch

### CODE REVIEW

**Comment by ronzulu** [comment]
Hi @KyleKlus thanks so much for doing this, especially with the detailed explanations on #870.

Hopefully I'll get a chance to review over the next couple of days. I kind of got excited about #748, and how much research has gone into the https://github.com/open-spaced-repetition/fsrs4anki algorithm.

Actually, would love your thoughts regarding #851 whilst you are "in the neighborhood"

Cheers
Ronny

**Comment by KyleKlus** [comment]
The new algorithm in #748 really looks nice and I think it could be a great benefit for the users. Maybe it could be first implemented as an opt-in feature so it can be tested first, how it behaves and feels.

I also had the same idea as mentioned in #851; I think it is particularly useful, so that one roughly knows how long the learning session might still take. And it would be very easy to implement it, as all data is already there and the ui has still some space for a subheading. 
The only thing to consider is a possible jump in the ui, if the root deck has cards and then after those are done it switches to the subdeck. The subheading could push the question section down if it appears, so I would pre allocate the space for the sub deck info, so it doesn't push anything down when it appears.

I would be happy to implement that while I am working on the other ui changes.

**Comment by ronzulu** [comment]
> The new algorithm in #748 really looks nice and I think it could be a great benefit for the users. Maybe it could be first implemented as an opt-in feature so it can be tested first, how it behaves and feels.

I was thinking the same 👍 

> I also had the same idea as mentioned in #851, but just for the normal review mode; I think it is particularly useful, so that one roughly knows how long the learning session might still take.

Good point, I hadn't thought of that

> The only thing to consider with the subheading is a possible jump in the ui, if the root deck has cards and then after those are done it switches to the subdeck. The subheading could push the question section down if it appears, so I would pre allocate the space for the sub deck info, so it doesn't push anything down when it appears.

I think I understand your point, personally it wouldn't bother me if occasionally after completing a card during the review, the next card is displayed at a slightly different vertical position because of the extra info. I only use Obsidian on the desktop so vertical space isn't a premium, as long as on a mobile the pre allocated space doesn't seem wasted when it's blank.

> I would be happy to implement that after I've made the other ui changes in #872 😊.

Awesome!

**Comment by KyleKlus** [comment]
Hi again, I am done now with all the ui improvements I wanted to do in #872. The PR for the this would build upon this unmerged PR, as I finalized/improved the view switching a bit there 🚀 [Link to the changes](https://github.com/KyleKlus/obsidian-spaced-repetition/tree/ui-overhaul).

So would the reviewing process be easier if I wait until this one is finished or would it be better, if I close this one and open another one with all the ui changes at once?

I just don't want to overwhelm the review process with too many changes 😇.

**Comment by ronzulu** [comment]
Hi @KyleKlus, I like your idea to make the UI changes under a separate PR. That way this PR is purely for the refactoring. Cheers

File: docs/changelog.md [hunk 1]
```diff
@@ -17,6 +17,8 @@ Generated by [`auto-changelog`](https://github.com/CookPete/auto-changelog).
 -   Improved how spacing between the buttons was done
```
Comment (ronzulu) [inline docs/changelog.md:17]: Nice detailed change log.
On this project the convention seems to be simply have the name, and issue or PR link.
Perhaps move the details to the PR description.

File: src/gui/DecksListView.tsx [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+import {
+    DeckStats,
+    IFlashcardReviewSequencer as IFlashcardReviewSequencer,
+} from "src/FlashcardReviewSequencer";
+import { TopicPath } from "src/TopicPath";
+import { FlashcardModalMode } from "./flashcard-modal";
+
+export class DecksListView {
+    public plugin: SRPlugin;
+    public mode: FlashcardModalMode;
```
Comment (ronzulu) [inline src/gui/DecksListView.tsx:19]: I don't think this is needed as it is always set to FlashcardModalMode.DecksList

File: docs/changelog.md (hunk 1 above)
Comment (KyleKlus) [reply docs/changelog.md]: I moved this now into the PR.
(Sry, that it took me a while, but I had to write some uni exams this week)

File: src/gui/DecksListView.tsx (hunk 2 above)
Comment (KyleKlus) [reply src/gui/DecksListView.tsx]: I removed this now from the class.

**Comment by KyleKlus** [comment]
I think, I fixed now all the warnings. Sry, that I didn't removed the unused variables; I don't know why I forgot that 😅.

**Review by st3v3nmw — APPROVED** [review]

**Comment by st3v3nmw** [comment]
LGTM, thanks!

**Pull request merged**